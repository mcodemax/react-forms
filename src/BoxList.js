import React, { useState } from "react";
import "./BoxList.css";
import Box from "./Box";
import NewBoxForm from "./NewBoxForm";
import { v4 as uuid } from 'uuid';

const dummyBoxes = [
    {height:"50px", width:"50px", backgroundColor:"#555", boxId: uuid()}, 
    {height:"50px", width:"50px", backgroundColor:"#555", boxId: uuid()}, 
    {height:"50px", width:"50px", backgroundColor:"#555", boxId: uuid()}, 
    {height:"50px", width:"50px", backgroundColor:"#555", boxId: uuid()}, 
    {height:"50px", width:"50px", backgroundColor:"#555", boxId: uuid()}]

console.log(dummyBoxes)

const BoxList = () => {
    //an ele in boxes is an obj:{height, width, backgroundColor}
    const [boxes, setBoxes] = useState([]); 
    
    //Donâ€™t change the state in the main body of the component. otherwise you get unlimited 
    // re-renders ask why

    /**func to remove box components */
    const removeBox = (boxId) => {
        //pass removeBox to Box component, then remove an instance of Box
        const newBoxes = boxes.filter( (box) => box.boxId !== boxId )
        setBoxes(newBoxes);
    };

    /**func to add box components */
    const addBox = (box) => {
        //add {height: , width: , backgroundColor: , id: } to boxes
        //when form submitted
        
        const newBoxes = [...boxes, box];
        setBoxes(newBoxes);
    };

    return (
        <>
            <div className="BoxList">
                {boxes.map( (box, ind) => {
                    return (
                        <Box key={ind} 
                        {...boxes[ind]} removeBox={removeBox} boxId={box.boxId}/>)
                })}
                <NewBoxForm addBox={addBox}/>
            </div>
        </>
    )
};

export default BoxList;


/*

func(
    //something has 5 elements
    something.map(
        //adds 1 to the state variable
    )
)
*/