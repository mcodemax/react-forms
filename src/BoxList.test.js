import React from "react";
import { render, fireEvent } from "@testing-library/react"; //but none is installed. You must install peer dependencies yourself. npm warning after installing react testing library
import BoxList from "./BoxList";
import { v4 as uuid } from 'uuid';

const dummyBoxes = [
    {height:"50px", width:"50px", backgroundColor:"#555", boxId: uuid()}, 
    {height:"50px", width:"50px", backgroundColor:"#555", boxId: uuid()}, 
    {height:"50px", width:"50px", backgroundColor:"#555", boxId: uuid()}, 
    {height:"50px", width:"50px", backgroundColor:"#555", boxId: uuid()}, 
    {height:"50px", width:"50px", backgroundColor:"#555", boxId: uuid()}]

it("should render", () => {
  render(<BoxList/>)
});

it("matches snapshot", () => {
    const obj = {
                  //get a boxList to render //maybe force dummyBoxes in BoxList somehow
                };
    const { asFragment } = render(<BoxList {...obj}/>);
    //maybe shove a mock f() in there so testing works
    //or alter BoxList to accept a preform'd box array so you don't have to worry about it
    expect(asFragment()).toMatchSnapshot();
});
