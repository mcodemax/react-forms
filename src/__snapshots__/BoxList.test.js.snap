// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="BoxList"
  >
    <form
      class="NewBoxForm"
    >
      <label
        for="width"
      >
        Width
      </label>
      <br />
      <input
        id="width"
        name="width"
        type="text"
      />
      <br />
      <label
        for="height"
      >
        Height
      </label>
      <br />
      <input
        id="height"
        name="height"
        type="text"
      />
      <br />
      <label
        for="backgroundColor"
      >
        Background Color of Box
      </label>
      <br />
      <input
        id="backgroundColor"
        name="backgroundColor"
        type="text"
      />
      <br />
      <input
        id="submit"
        name="submit"
        type="submit"
      />
    </form>
  </div>
</DocumentFragment>
`;
